include/tvm/runtime/registry.h

```diff
@@ -52,6 +52,45 @@
 namespace tvm {
 namespace runtime {
 
+/*!
+ * \brief Check if signals have been sent to the process and if so
+ *  invoke the registered signal handler in the frontend environment.
+ *
+ *  When runnning TVM in another langugage(python), the signal handler
+ *  may not be immediately executed, but instead the signal is marked
+ *  in the interpreter state(to ensure non-blocking of the signal handler).
+ *
+ *  This function can be explicitly invoked to check the cached signal
+ *  and run the related processing if a signal is marked.
+ *
+ *  Invoke this function periodically in a long running C++ function
+ *  to check if KeyboardInterrupt happens in a python execution environment.
+ *
+ *  Not inserting this function will not cause any correctness
+ *  issue, but will delay the KeyboardInterrupt until the function returns
+ *  to the python side. So this function is not needed in most API
+ *  functions can finish quickly in a reasonable amount of time.
+ *
+ * \code
+ *
+ * int check_signal_every_kiter = 10;
+ *
+ * for (int iter = 0; iter < very_large_number; ++iter) {
+ *   if (iter % check_signal_every_kiter == 0) {
+ *     tvm::runtime::EnvCheckSignals();
+ *   }
+ *   // do work here
+ * }
+ *
+ * \endcode
+ *
+ * \note This function is a nop when no signal checking function is registered.
+ *
+ * \throws This function throws approperiate exception if an error happens,
```

Comment:
This is not accurate. The error only happens when signal handler throws an error, will update

Reply:
ah you are correct, my mistake. however, I don't agree with the implementation here. `PyErr_CheckSignals` states:

>It checks whether a signal has been sent to the processes and if so, invokes the corresponding signal handler. If the signal module is supported, this can invoke a signal handler written in Python. In all cases, the default effect for SIGINT is to raise the KeyboardInterrupt exception. 

If, in Python, I do: 
```
old_sighandler = None
def cleanup_temp_files(*args):
  shutil.rmtree("/tmp/foo")
  old_sighandler(*args)

old_sighandler = signal.signal(signal.SIGINT, cleanup_temp_files)
```

which is perfectly legal, but `shutil.rmtree` throws an exception, you will mask that exception here with KeyboardInterrupt. 

Instead, it seems we need to throw a special C++-side exception which is caught in TVMFuncCall and then return normally to Python as if no error occurred (perhaps first asserting that `PyErr_Occurred() != nullptr`). this allows whatever exception was raised by the signal handler to propagate normally with correct traceback information. if for some reason `PyErr_Occurred() == nullptr` at this time, we could continue to raise a TVM-side error indicating that a signal handler interrupted the call. Perhaps the return value of TVMFuncCall should be modified to account for this scenario.

In addition, I think it makes sense to push the handling of this into the FFI. Different frontends may handle this differently--for instance, other frontends may not be capable of executing the signal handler from a callback function such as `PyErr_CheckSignals`, so it seems like it makes sense to me merely to treat such a function as "should I continue?" and return an indicative error code when that function says no.

Reply:
The main complexity of pushing the handling to FFI layer is that we will need to introduce specific compiled functions in the FFI layer(via cython and others)  complicates the implementation(we cannot use python to implement it because the handling have to go through the C API, which is doable but introduces complication to implement such function in cython without any support from the python land). 

With additional effort we might be able to recover the exact exception raised. Given KeyboardInterrupt would be the most common one. We could change that to a generic exception indicating that an error happens during signal handling.

I have also tried to propagate the exact error to the python side would be harder because it involves complications in the python calling boundary(we cannot do so in the python land and it have to be done in a C++ land). Directly normal return could raise an SystemError that also shows the original error message.

Finally, back to the generic function. There are two ways to think about the handling. Given the only usecase so far is python I trid to avoid generalizing and centralizes the implementation in the runtime(to reduce indirection). We can refactor once we see more needs and given there is no API change such refactor should be easy

Reply:
> The main complexity of pushing the handling to FFI layer is that we will need to introduce specific compiled functions in the FFI layer

In this case, the functions are designed to be used with the FFI layer. So I'm not sure I see a problem with moving the use of the functions closer to the place it's intended to live.

In this case, the complexity is actually the same, because you remove the need to expose PyErr_Clear() to TVM and instead replace its use with asserting `PyError_Occurred() != nullptr`. The complexity in the TVM runtime is then reduced to that around the single frontend-agnostic function which asks "should I continue?"

Could you clarify about "I have also tried to propagate the exact error to the python side?" Theoretically, you should not need to propagate any error here--Python should take care of that.

I'm okay with how we treat the generic function now, though I do think that if there is just one, it may be a bit overcomplicated to introduce a registry. A single PackedFunc should suffice here.